Mathlib/Analysis/InnerProductSpace/Positive.lean

```diff
@@ -206,6 +206,98 @@ end PartialOrder
 
 end ContinuousLinearMap
 
+/-- a star projection is an idempotent and self-adjoint element -/
+structure IsStarProjection {M : Type*} [Mul M] [Star M] (p : M) : Prop where
+  isIdempotentElem : IsIdempotentElem p
+  isSelfAdjoint : IsSelfAdjoint p
+
+/-- a star projection is non-negative in a star-ordered ring -/
+theorem IsStarProjection.nonneg {M : Type*} [NonUnitalSemiring M] [PartialOrder M]
+  [StarRing M] [StarOrderedRing M] {p : M}
+  (hp : IsStarProjection p) : 0 ≤ p :=
+by
+  rw [← hp.isIdempotentElem]
+  nth_rw 1 [← hp.isSelfAdjoint]
+  exact star_mul_self_nonneg p
+
+-- this will be removed in the future when we have `StarOrderedRing (E →L[𝕜] E)`
+-- as this is just `IsStarProjection.nonneg`
+/-- a star projection operator is positive -/
+theorem isStarProjection_isPositive [CompleteSpace E] {p : E →L[𝕜] E}
+  (hp : IsStarProjection p) : 0 ≤ p :=
+by
+  rw [ContinuousLinearMap.nonneg_iff_isPositive]
+  refine ⟨hp.isSelfAdjoint, ?_⟩
+  rw [← hp.isIdempotentElem.eq]
+  simp_rw [reApplyInnerSelf_apply, ContinuousLinearMap.mul_apply]
+  intro x
+  simp_rw [← ContinuousLinearMap.adjoint_inner_right _ _ x, isSelfAdjoint_iff'.mp hp.isSelfAdjoint]
+  exact inner_self_nonneg
+
+theorem ContinuousLinearMap.ker_eq_ortho_adjoint_range {V W : Type*}
+  [NormedAddCommGroup V] [NormedAddCommGroup W]
+  [InnerProductSpace 𝕜 V] [InnerProductSpace 𝕜 W]
+  [CompleteSpace V] [CompleteSpace W] (T : V →L[𝕜] W) :
+    ker T = (range (adjoint T))ᗮ := by
+  ext
+  simp_rw [Submodule.mem_orthogonal, mem_range, mem_ker,
+    forall_exists_index, forall_apply_eq_imp_iff,
+    ContinuousLinearMap.adjoint_inner_left]
+  exact ⟨fun h => by simp_rw [h, inner_zero_right, forall_const],
+    fun h => inner_self_eq_zero.mp (h _)⟩
+
+theorem IsStarProjection.ker_eq_ortho_adjoint_range
+  [CompleteSpace E] {p : E →L[𝕜] E} (hp : IsStarProjection p) :
+    LinearMap.ker p = (LinearMap.range p)ᗮ :=
```

Comment:
Maybe you can just rewrite this into a generic `IsSymmetric` or `IsSelfAdjoint` lemma for continuous linear maps?